### PR TITLE
quic: add QUIC downstream connection close error stats.

### PR DESCRIPTION
**Pull request by RenjieTang**
Signed-off-by: Renjie Tang <renjietang@google.com>

Commit Message: Introduce QuicStatNames and add Quic connection close stats for downstream connections.
Risk Level: Low
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by RenjieTang** [comment]
/assign @alyssawilk 
Hi Alyssa, would you mind taking a first pass review on this PR?
I think currently this change works well with H3 enabled. But I’m having trouble making it work without H3.
Since QuicStats exists as class members and is passed down in constructors, is there any good way to avoid the build dependency when H3 is disabled? I could imagine adding extra constructors, but not sure if that’s the right approach.

**Review by alyssawilk — COMMENTED** [review]
Hm.

At first pass my thought was that all our existing http3 stats are rooted in the HCM, and it'd be nice to be consistent here.
I'm not sure if that's the right thing here, because of the L4/L7 thing, but given that for TCP we don't have the stats in the connection, but we have the tcp network filter or the HCM do the accounting, I think a utility we can reuse when we add other protocols plus wiring it up in the HCM might be the right thing.  Tagging @jmarantz for a second opinion as stats are more his area.

If we go that route, I think we could stash the close details in connection level stream info as one of those generic extension points, and pull it out in the HCM if it's present.  We've already got a few h3 stats in the HCM which we don't bother compiling out, so as long as we added lazily I think it wouldn't be a problem.

**Comment by RenjieTang** [comment]
I saw that HttpConnectionManagerConfig owns http3_codec_stats_ and passes it down to QuicHttpServerConnectionImpl.
Maybe we can have QuicStats to follow the same pattern?

**Comment by jmarantz** [comment]
clang-tidy error needs attention:
```
2021-05-19T22:37:06.0307772Z Writing fixes to clang-tidy-fixes.yaml ...
2021-05-19T22:37:06.0815884Z clang-tidy check failed, potentially fixed by clang-apply-replacements:
2021-05-19T22:37:06.0853537Z Diagnostics:
2021-05-19T22:37:06.0854670Z - BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
2021-05-19T22:37:06.0855193Z   DiagnosticMessage:
2021-05-19T22:37:06.0855548Z     FileOffset: 619
2021-05-19T22:37:06.0856073Z     FilePath: source/common/quic/quic_stats.h
2021-05-19T22:37:06.0856758Z     Message: invalid case style for function 'ConnectionCloseStatName'
2021-05-19T22:37:06.0857206Z     Replacements:
2021-05-19T22:37:06.0857930Z     - {FilePath: source/common/quic/quic_stats.h, Length: 23, Offset: 619, ReplacementText: connectionCloseStatName}
2021-05-19T22:37:06.0858685Z   DiagnosticName: readability-identifier-naming
2021-05-19T22:37:06.0859093Z   Level: Warning
2021-05-19T22:37:06.0859745Z - BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
2021-05-19T22:37:06.0864759Z   DiagnosticMessage:
2021-05-19T22:37:06.0865156Z     FileOffset: 3737
2021-05-19T22:37:06.0865566Z     FilePath: source/common/quic/active_quic_listener.h
2021-05-19T22:37:06.0866392Z     Message: invalid case style for parameter 'quic_stats_'
2021-05-19T22:37:06.0866825Z     Replacements:
2021-05-19T22:37:06.0867492Z     - {FilePath: source/common/quic/active_quic_listener.h, Length: 11, Offset: 3737,
2021-05-19T22:37:06.0868006Z       ReplacementText: quic_stats}
2021-05-19T22:37:06.0868748Z   DiagnosticName: readability-identifier-naming
2021-05-19T22:37:06.0869157Z   Level: Warning
2021-05-19T22:37:06.0869634Z MainSourceFile: ''
2021-05-19T22:37:06.3876930Z ##[error]Bash exited with code '1'.
2021-05-19T22:37:06.4036353Z ##[section]Finishing: Run CI script
```

**Comment by RenjieTang** [comment]
> clang-tidy error needs attention:
> 
> ```
> 2021-05-19T22:37:06.0307772Z Writing fixes to clang-tidy-fixes.yaml ...
> 2021-05-19T22:37:06.0815884Z clang-tidy check failed, potentially fixed by clang-apply-replacements:
> 2021-05-19T22:37:06.0853537Z Diagnostics:
> 2021-05-19T22:37:06.0854670Z - BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
> 2021-05-19T22:37:06.0855193Z   DiagnosticMessage:
> 2021-05-19T22:37:06.0855548Z     FileOffset: 619
> 2021-05-19T22:37:06.0856073Z     FilePath: source/common/quic/quic_stats.h
> 2021-05-19T22:37:06.0856758Z     Message: invalid case style for function 'ConnectionCloseStatName'
> 2021-05-19T22:37:06.0857206Z     Replacements:
> 2021-05-19T22:37:06.0857930Z     - {FilePath: source/common/quic/quic_stats.h, Length: 23, Offset: 619, ReplacementText: connectionCloseStatName}
> 2021-05-19T22:37:06.0858685Z   DiagnosticName: readability-identifier-naming
> 2021-05-19T22:37:06.0859093Z   Level: Warning
> 2021-05-19T22:37:06.0859745Z - BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
> 2021-05-19T22:37:06.0864759Z   DiagnosticMessage:
> 2021-05-19T22:37:06.0865156Z     FileOffset: 3737
> 2021-05-19T22:37:06.0865566Z     FilePath: source/common/quic/active_quic_listener.h
> 2021-05-19T22:37:06.0866392Z     Message: invalid case style for parameter 'quic_stats_'
> 2021-05-19T22:37:06.0866825Z     Replacements:
> 2021-05-19T22:37:06.0867492Z     - {FilePath: source/common/quic/active_quic_listener.h, Length: 11, Offset: 3737,
> 2021-05-19T22:37:06.0868006Z       ReplacementText: quic_stats}
> 2021-05-19T22:37:06.0868748Z   DiagnosticName: readability-identifier-naming
> 2021-05-19T22:37:06.0869157Z   Level: Warning
> 2021-05-19T22:37:06.0869634Z MainSourceFile: ''
> 2021-05-19T22:37:06.3876930Z ##[error]Bash exited with code '1'.
> 2021-05-19T22:37:06.4036353Z ##[section]Finishing: Run CI script
> ```

done

File: source/common/quic/active_quic_listener.cc
Comment (jmarantz) [inline source/common/quic/active_quic_listener.cc:40]: I'd suggest bundling QuicStats and quic::QuicConfig into a class Quic::Context, which can parallel Http::Context in most ways.

That will cut down by one on the arg-count in a few interfaces in this PR, and give you other good places to put stuff that may need to be shared across workers.

Moreover, `Quic::Context` (or maybe `OptRef<Quic::Context>`) could then be returned from `Server::FactoryContext`, also parallel to Http::Context and Grpc::Context, cutting down on how much plumbing through needs to happen for those individually.

File: source/common/quic/envoy_quic_dispatcher.cc
Comment (jmarantz) [inline source/common/quic/envoy_quic_dispatcher.cc:20]: per above, pass through QuicContext rather than QuicStats and QuicConfig separately.

File: source/common/quic/envoy_quic_dispatcher.cc
Comment (jmarantz) [inline source/common/quic/envoy_quic_dispatcher.cc:49]: I think you should just make QuicStats unconditional at this level. I don't see the problem with instantiating QuicStats in the places below where you are currently passing nullptr.

**Review by jmarantz — COMMENTED** [review]
This generally looks great! I just have a few comments about making the stats contexts parallel those for HTTP and GRPC for increased consistency and reduced number of args at various interface-points.

**Review by RenjieTang — COMMENTED** [review]
Thanks for the review Joshua! Glad I got most of the QuicStats class right.
I left some in-place comments on the ownership and plumbing-through of QuicStats.
In general, I think it might not be the best choice to have the Server::InstanceImpl own QuicStats and pass it down all the way.

Alyssa mentioned in comments that a lot of existing QUIC stats live in HttpConnectionManagerConfig. I tried placing QuicStats there and found things to be a lot cleaner. The demo PR can be found at https://github.com/envoyproxy/envoy/pull/16625. Let me know what you think on that.

File: source/common/quic/active_quic_listener.cc
Comment (jmarantz) [reply source/common/quic/active_quic_listener.cc:40]: On this point, to match the pattern in the other Contexts is to forward-declare QuicStats without defining an interface for it.

So you'd have
```
namespace quic { class QuicConfig; }
namespace Envoy {
namespace Quic {
class QuicStats;
class Context {
 public:
  virtual QuicStats& stats() PURE:
  virtual quic::QuicConfig& config() PURE;
};
...
```
So you don't need to expose the error-code type or other types at the interface.

**Review by jmarantz — COMMENTED** [review]
I'll check out the other PR a bit later.

**Comment by jmarantz** [comment]
> In general, I think it might not be the best choice to have the Server::InstanceImpl own QuicStats and pass it down all the way.

Can you explain the reasoning?

File: source/server/server.cc
Comment (jmarantz) [reply source/server/server.cc:108]: Is there a configuration option for Quic? Or you have to make the decision about whether to support it at compile-time?

**Comment by RenjieTang** [comment]
> > In general, I think it might not be the best choice to have the Server::InstanceImpl own QuicStats and pass it down all the way.
> 
> Can you explain the reasoning?

I was mainly concerned that having QuicContext accessible in Server::Instance will require getter implementation in subclasses that don't really need it, for example Server::ValidationInstance. If having QuicStats in HCM works just as well, I'd prefer to go that route so that the code is cleaner.

File: source/server/server.cc
Comment (RenjieTang) [reply source/server/server.cc:108]: The QUIC support is decided at compile time and is controlled by --//bazel:http3=True/False

File: source/server/listener_impl.cc
Comment (jmarantz) [inline source/server/listener_impl.cc:404]: change to `#ifdef` for consistency?

File: source/server/listener_impl.h
Comment (jmarantz) [inline source/server/listener_impl.h:235]: revert blank line.

File: source/server/listener_impl.h
Comment (jmarantz) [inline source/server/listener_impl.h:339]: Rather than having this OptRef flow can we set this in the constructor from the ListenerManagerFactory that's passed in?

File: source/server/listener_manager_impl.cc
Comment (jmarantz) [inline source/server/listener_manager_impl.cc:439]: As above I think ListenerImpl can just grab the `QuicStatNames&` object from the manager.

**Review by jmarantz — COMMENTED** [review]
this is looking much better. I think we can just get rid of the one OptRef.

File: source/server/listener_manager_impl.cc
Comment (jmarantz) [reply source/server/listener_manager_impl.cc:439]: Or...if that involves changes to the base class you'd prefer not to do, you can just pass in the stats to the constructor rather than setting it after construction.

That just means you need to ifdef the arg-list.

Comment (RenjieTang) [reply source/server/listener_manager_impl.cc]: Fortunately ListenerImpl takes ListenerManagerImpl in its constructors. So I added getter in ListenerManagerImpl and let ListenerImpl grab it during constructions.

**Review by jmarantz — APPROVED** [review]
@envoyproxy/senior-maintainers 

thanks. for pushing this all the way through. nice job!

File: source/common/quic/quic_stat_names.cc
Comment (alyssawilk) [inline source/common/quic/quic_stat_names.cc:12]: as discussed let's do the server side MVP too

File: source/common/quic/quic_stat_names.cc
Comment (alyssawilk) [inline source/common/quic/quic_stat_names.cc:18]: maybe comment an example counter name?

File: test/common/quic/quic_stat_names_test.cc
Comment (alyssawilk) [inline test/common/quic/quic_stat_names_test.cc:21]: We should doc up the new stats somewhere under docs/

File: source/common/quic/quic_stat_names.h
Comment (alyssawilk) [inline source/common/quic/quic_stat_names.h:30]: does this PR do upstream stats?  Should we have a TODO to wire those up?

File: test/common/quic/quic_stat_names_test.cc
Comment (alyssawilk) [inline test/common/quic/quic_stat_names_test.cc:25]: can we add checks to one of the quic http integration tests in test/integration/quic_http_integration_test.cc

**Review by alyssawilk — COMMENTED** [review]
Awesome - these are going to be super useful!

File: source/common/quic/quic_stat_names.cc
Comment (RenjieTang) [reply source/common/quic/quic_stat_names.cc:15]: done

File: source/common/quic/quic_stat_names.cc
Comment (RenjieTang) [reply source/common/quic/quic_stat_names.cc:21]: done adding comments in the method declaration.

File: source/common/quic/quic_stat_names.h
Comment (RenjieTang) [reply source/common/quic/quic_stat_names.h:33]: Currently this is only for downstream since our downstream customers have a tighter timeline. Added TODO to plumb it through upstreams.

File: test/common/quic/quic_stat_names_test.cc
Comment (RenjieTang) [reply test/common/quic/quic_stat_names_test.cc:21]: I added a new section in "Per listener stat" for Http3 stats.
I think it's a better place to document these new stats compared to codec_stats from the perspective of stats scopes.

File: docs/root/configuration/http/http_conn_man/stats.rst
Comment (alyssawilk) [inline docs/root/configuration/http/http_conn_man/stats.rst:118]: If we're doccing up listener stats, won't direction always be downstream?

And actually I hate to catch this now, but probably rather than self/peer it should be tx/rx for consistency with other Envoy stats.  I find self and peer more intuitive (maybe 'cause that's what I'm used to) but I think most of the other stats (like resets tx_reset/rx_reset) are documented that way with tx for things sent to the peer and rx being codes received from the peer.

**Review by alyssawilk — COMMENTED** [review]
One naming nit, and I think this is good to go from my point of view.

Adding Matt for a non-google pass since stats are forever :-P

File: docs/root/configuration/http/http_conn_man/stats.rst
Comment (RenjieTang) [reply docs/root/configuration/http/http_conn_man/stats.rst]: You are right. The listener stats should be all downstreams.

I haven't figured out the details on upstream stats. So I will update the doc again when I add them.

Done changing names to tx/rx.

**Comment by mattklein123** [comment]
Check CI?

/wait

**Review by alyssawilk — APPROVED** [review]

**Comment by RenjieTang** [comment]
@mattklein123  The flaky tests are now fixed. PTAL.

File: docs/root/configuration/http/http_conn_man/stats.rst
Comment (mattklein123) [inline docs/root/configuration/http/http_conn_man/stats.rst:92]: nit: s/Http/HTTP here and below

File: source/common/quic/quic_stat_names.cc
Comment (mattklein123) [inline source/common/quic/quic_stat_names.cc:7]: From a quick read through it's not clear to me that this won't create an unbounded number of stats controlled by downstream (for exampled if the name somehow has the code number in it and the client controls the codes). This is probably not the case, but can you add more comments about why this is safe for use with uncontrolled peers?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with some small comments.

/wait

File: source/common/quic/quic_stat_names.cc
Comment (RenjieTang) [reply source/common/quic/quic_stat_names.cc:7]: Great point!
I initially had an assertion in QuicStatNames::connectionCloseStatName() to make sure we don't create unlimited amount of stats.
But after some investigation, I found that if the connection close is initiated from peers, the error_code is parsed from the wire and no enum range checking is done. So an assertion here is too strong and malicious clients might be able to attack Envoy by sending out-of-range error codes.

Instead, I now added a check in QuicStatNames::chargeQuicConnectionCloseStats() to ignore out-of-range error_codes and log a warning.

I will follow up to dig deeper into the quiche code and see if bad error code handling can be done earlier.

File: source/common/quic/quic_stat_names.cc
Comment (mattklein123) [inline source/common/quic/quic_stat_names.cc:32]: Can you add a test case for this? Also, I think warning here can log spam. I would recommend making this debug level and potentially also adding a stat for unknown error code or something like that. Thank you.

/wait

Comment (RenjieTang) [reply source/common/quic/quic_stat_names.cc]: done.
I utilized quic::QUIC_LAST_ERROR_CODE as a stat for all out of range codes.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
